### ISSUE TITLE
PHP 8 Support

### ISSUE THREAD

**Issue by vigneshgurusamy** [phpseclib/phpseclib#1548]
Hi,

Will version **2.x** support PHP 8 or do we need to use version **3.x** for PHP 8 support?

I heard that the version **3.x** is moving to throw exceptions instead of errors. It will be helpful for us, if we know the release timeline for version **3.x**.

Thanks in advance.

**Comment by terrafrost** [comment]
> Will version 2.x support PHP 8 or do we need to use version 3.x for PHP 8 support?

That's the plan. I just need to get PHP 8 working with Travis CI. Travis CI didn't have PHP 8 support yesterday but it looks like it does as of today.

> I heard that the version 3.x is moving to throw exceptions instead of errors. It will be helpful for us, if we know the release timeline for version 3.x.

As soon as I get PHP 8 working on Travis CI. Looks like they just added PHP 8 support today. The last major holdup was the documentation website (https://phpseclib.com/), which went live a few days ago.

**Comment by terrafrost** [comment]
Well it looks like Travis CI no longer supports PHP 8. It was working Friday morning but by Friday afternoon it wasn't, again. Getting the same error I was getting before:

```
$ phpenv global 8.0 2>/dev/null
8.0 is not pre-installed; installing
Downloading archive: https://storage.googleapis.com/travis-ci-language-archives/php/binaries/ubuntu/18.04/x86_64/php-8.0.tar.bz2
$ curl -sSf --retry 5 -o archive.tar.bz2 $archive_url && tar xjf archive.tar.bz2 --directory /
$ phpenv global 8.0
rbenv: version `8.0' not installed
The command "phpenv global 8.0" failed and exited with 1 during .
Your build has been stopped.
```
Quoting https://travis-ci.community/t/php-8-0-missing/10132/17 , "_There is an issue with the `8.0` archive, so that will be addressed net week_"

Before that happened what I had been doing was updating the unit tests for the 1.0 branch to work with PHPUnit 9. So right now phpseclib is unit tested with, at latest, PHPUnit 6.0, but that doesn't work on PHP 8. I could fork PHPUnit 6.0, I suppose, and make it work on PHP 8 but I'd kinda prefer to do PHPUnit 9. I can't just update the composer.json, however, because there are BC breaking changes between the two versions of PHPUnit:

- PHPUnit 9 expects class names to be of form SFTPLargeFileTest vs Functional_Net_SFTPLargeFileTest. The former is consistent with PSR4 vs the latter, which is PSR0
- `setUpBeforeClass()`, `setUp()` and `tearDown()` need `: void` appended to them (void return types were introduced in PHP 7.1)
- `assertInternalType('string', ...)` is now `assertIsString(...)`
- `assertContains`  with two strings is now `assertStringContainsString`
- etc

After making these changes I found an issue in ANSI.php relating to the fact that in PHP 7.4 `'' == 0` whereas in PHP 8.0 that is no longer the case (see https://wiki.php.net/rfc/string_to_number_comparison).

As of Thursday night (CST) the unit tests were still failing and I'm not sure why. Without [GNU parallel](https://en.wikipedia.org/wiki/GNU_parallel) the tests timed out after [97% of them had been ran](https://travis-ci.com/github/terrafrost/phpseclib/builds/206502684). _With_ GNU parallel the tests finish but they're still showing up as failing in Travis CI. And yet the words "fail" and "error" do not appear in [the output](https://travis-ci.com/github/terrafrost/phpseclib/builds/206500857). So idk wtf is up with that. My plan is to do a binary search of the unit tests until I find the one that's causing the failure since examining the output is apparently a fruitless endeavor. But I guess I'll try this next week when Travis CI is working with PHP 8 again!

Overall, tho, phpseclib does seem to work without issue on PHP 8. Like I said, outside of ANSI.php, 97%+ of the unit tests were passing.

**Comment by vigneshgurusamy** [comment]
I really appreciate your effort 😊... will wait for your update...

**Comment by terrafrost** [comment]
## With GNU parallels
I've gotten this working in PHP 8:

https://github.com/terrafrost/phpseclib/commit/c36fd67c2901625aaec5cc0fb9046326c45c0ea2
https://travis-ci.com/github/terrafrost/phpseclib/builds/207377997

Apparently the reason it was failing on GNU parallels was that the class names in the Unit/Crypt/Hash directory and in Unit/Crypt/RSA hadn't been updated. And then, for Unit/Crypt/Hash/SHA256_96Test...  that test extends SHA256Test but I guess that test isn't being autoloaded so I had to add a `require_once`.

idk why GNU parallels wasn't showing that. Honestly, that's one thing about GNU parallels I really dislike - it sometimes hides stuff.

## Without GNU parallels
As for why this was hanging when GNU parallels _wasn't_ in use...  idk. We can see that it does here:

https://travis-ci.com/github/terrafrost/phpseclib/builds/207392392

I deleted the Unit tests, keeping only the Functional tests, and it worked:

https://travis-ci.com/github/terrafrost/phpseclib/builds/207391373

I then deleted the Functional tests, keeping only the Unit tests, and it worked:

https://travis-ci.com/github/terrafrost/phpseclib/builds/207391699

So if both directories work without issue, in isolation, why do they timeout when combined? idk. Still trying to figure that out...

## Current Game Plan
My current game plan is to make each unit test contain two classes - one PSR0 style one and one PSR4 style one that'll extend the PSR0 style one.

I'll use sed to add `: void` after the relevant classes, I'll add `assertIsString()` to PhpseclibTestCase to emulate it when it isn't present in PHPUnit\Framework\TestCase, etc.

I also want to know why it's not working when GNU parallels isn't being used.

**Comment by terrafrost** [comment]
I figured out why it was failing without GNU parallel.

So for the Hash unit tests it was doing `define('CRYPT_HASH_MODE', CRYPT_HASH_MODE_INTERNAL)` in `setUpBeforeClass()`. Without GNU Parallel, the Hash unit tests were running before `SFTPLargeFileTest`, which attempts to upload a 128MB file, so it was using the slowest possible mode for creating sha512 hashes, so it was timing out. _With_ GNU Parallel, however, `SFTPLargeFileTest` was being ran independently of the Hash unit tests and so, in-so-far as `SFTPLargeFileTest` was concerned, CRYPT_HASH_MODE was CRYPT_HASH_MODE_HASH.

Anyway, the fix was to do this:

https://github.com/terrafrost/phpseclib/commit/d25ea5a06066519159b10c52b9b8250cba14069b

Now that I've figured this out the next step is to start making it these unit tests can pass in both PHPUnit for PHP 5.x and in PHPUnit for 8.0, using the roadmap I discussed in my last post. I'll try to work on this this weekend!
